### ISSUE TITLE
Finding an input that maximizes a class for text classification

### ISSUE THREAD

**Issue by sebastianruder** [fchollet/keras#1923]
I would like to replicate @fchollet's example [here](http://blog.keras.io/how-convolutional-neural-networks-see-the-world.html) for text classification to obtain sequences that maximize the activations of specific filters or the activiations towards a certain output class.

I'm running on TensorFlow and have loaded the weights of a trained `Sequential` model with a `Convolution1D` layer.
Instead of `input_img` in the post, I set `input_seq = K.placeholder((MAX_LEN, MAX_FEATURES))`.
I access the output of a specified layer as in the example: `layer_output = layer_dict[layer_name].get_output()`.
I calculate the loss as `loss = K.mean(layer_output[:, filter_index, :])`.
However, computing the gradients with `K.gradients(loss, input_seq)` yields `[None]`.

Could you tell me what I'm doing wrong? Thank you so much for your help!

**Comment by valexandersaulys** [comment]
So I can't see your full code, but I believe you're passing through an input of two dimensions when Convolution1D expects three. See the documentation here for what I'm referring to http://keras.io/layers/convolutional/.

Thus your ```input_seq``` needs to be set to something like ```input_seq = K.placeholder((DIMENSION_1, DIMENSION_2, DIMENSION_3))``` 

I don't know the nature of your data, so I don't know if this is what you want or need.

**Comment by sebastianruder** [comment]
Thanks for the prompt reply! You're right. I somehow forgot that I need to feed the sequence directly into the convolutional layer.

I've now set `input_seq = K.placeholder((1, nb_filter, filter_length))`.

However, `K.gradients(loss, input_seq)` still returns `[None]`. 

For context, here is a simplified model that does what I want to do:
```
    model = Sequential()
    model.add(Embedding(max_features, embedding_size, input_length=max_len))
    model.add(Convolution1D(filter_length=filter_length, nb_filter=nb_filter))
    model.add(MaxPooling1D())
    model.add(Flatten())
    model.add(Dense(output_dim=num_classes))
    model.compile(loss='categorical_crossentropy', optimizer='adadelta')
```

**Comment by valexandersaulys** [comment]
Convolution1D expects ```(samples, steps, input_dim)``` as its 3d input. By giving an ```input_seq = K.placeholder((1, nb_filter, filter_length))`` Keras is getting a sample size of 1 and thus doing unexpected things. You should have multiple samples in your dataset for training.

Methinks you're using the wrong tool for the job here. You might be better off with a Dense layer or something that accepts 2D data.

**Comment by sebastianruder** [comment]
For training, sure, you pass in multiple samples.
Note that in this case, the model is already trained and my goal is to find _one_ sample that maximizes the activations of one layer or towards one output class, so that I can then map this back to a text representation.

**Comment by satishjasthi** [comment]
I'm also facing the same error can anyone help

**Comment by adekuar** [comment]
You are getting gradient as None because the embedding layer in keras are not differentiable. I have been trying to achieve the same that you tried, couldn't find a solution so far.
